File: ros2_ws/src/cpp_action/src/fibonacci_action_client.cpp
#include "cpp_action/fibonacci_action_client.hpp"

#include <chrono>
#include <sstream>

namespace learning_ros2::cpp_action {

FibonacciActionClient::FibonacciActionClient(const rclcpp::NodeOptions& options)
    : rclcpp::Node("fibonacci_action_client", options),
      cancel_after_(-1),
      feedback_count_(0) {
  client_ = rclcpp_action::create_client<Fibonacci>(this, "fibonacci");
}

void FibonacciActionClient::send_goal(int32_t order, int32_t cancel_after) {
  cancel_after_ = cancel_after;
  feedback_count_ = 0;

  if (!client_->wait_for_action_server(std::chrono::seconds(10))) {
    RCLCPP_ERROR(this->get_logger(),
                 "Action server not available after waiting");
    rclcpp::shutdown();
    return;
  }

  auto goal_msg = Fibonacci::Goal();
  goal_msg.order = order;

  RCLCPP_INFO(this->get_logger(), "Sending goal");

  auto send_goal_options = rclcpp_action::Client<Fibonacci>::SendGoalOptions();
  send_goal_options.goal_response_callback =
      std::bind(&FibonacciActionClient::goal_response_callback, this,
                std::placeholders::_1);
  send_goal_options.feedback_callback =
      std::bind(&FibonacciActionClient::feedback_callback, this,
                std::placeholders::_1, std::placeholders::_2);
  send_goal_options.result_callback = std::bind(
      &FibonacciActionClient::result_callback, this, std::placeholders::_1);

  client_->async_send_goal(goal_msg, send_goal_options);
}

void FibonacciActionClient::goal_response_callback(
    const GoalHandleFibonacci::SharedPtr& goal_handle) {
  if (!goal_handle) {
    RCLCPP_ERROR(this->get_logger(), "Goal was rejected by server");
    rclcpp::shutdown();
    return;
  }

  RCLCPP_INFO(this->get_logger(),
              "Goal accepted by server, waiting for result");
  goal_handle_ = goal_handle;
}

void FibonacciActionClient::feedback_callback(
    GoalHandleFibonacci::SharedPtr,
    const std::shared_ptr<const Fibonacci::Feedback> feedback) {
  feedback_count_++;
  std::stringstream ss;
  ss << "Received feedback #" << feedback_count_ << ": [";
  for (size_t i = 0; i < feedback->partial_sequence.size(); ++i) {
    ss << feedback->partial_sequence[i];
    if (i + 1 < feedback->partial_sequence.size()) {
      ss << ", ";
    }
  }
  ss << "]";
  RCLCPP_INFO(this->get_logger(), "%s", ss.str().c_str());

  if (cancel_after_ > 0 && feedback_count_ >= cancel_after_) {
    RCLCPP_INFO(this->get_logger(), "Requesting cancel...");
    client_->async_cancel_goal(goal_handle_);
  }
}

void FibonacciActionClient::result_callback(
    const GoalHandleFibonacci::WrappedResult& result) {
  switch (result.code) {
    case rclcpp_action::ResultCode::SUCCEEDED:
      RCLCPP_INFO(this->get_logger(), "Goal succeeded");
      break;
    case rclcpp_action::ResultCode::CANCELED:
      RCLCPP_INFO(this->get_logger(), "Goal canceled");
      break;
    case rclcpp_action::ResultCode::ABORTED:
      RCLCPP_ERROR(this->get_logger(), "Goal aborted");
      break;
    default:
      RCLCPP_ERROR(this->get_logger(), "Unknown result code");
      break;
  }

  std::stringstream ss;
  ss << "Result: [";
  for (size_t i = 0; i < result.result->sequence.size(); ++i) {
    ss << result.result->sequence[i];
    if (i + 1 < result.result->sequence.size()) {
      ss << ", ";
    }
  }
  ss << "]";
  RCLCPP_INFO(this->get_logger(), "%s", ss.str().c_str());

  rclcpp::shutdown();
}

}  // namespace learning_ros2::cpp_action

File: ros2_ws/src/py_action/py_action/fibonacci_action_client.py
from typing import Optional

import rclpy
from rclpy.action import ActionClient
from rclpy.action.client import ClientGoalHandle
from rclpy.node import Node
from tutorial_interfaces.action import Fibonacci


class FibonacciActionClient(Node):
    def __init__(self) -> None:
        super().__init__("fibonacci_action_client")
        self.action_client = ActionClient(self, Fibonacci, "fibonacci")
        self.goal_handle: Optional[ClientGoalHandle] = None
        self.feedback_count: int = 0
        self.cancle_after: int = -1

    def send_goal(self, order: int, cancel_after: int = -1) -> None:
        goal_msg = Fibonacci.Goal()
        goal_msg.order = order

        self.cancle_after = cancel_after
        self.feedback_count = 0

        self.action_client.wait_for_server()

        self.send_goal_future = self.action_client.send_goal_async(
            goal_msg, feedback_callback=self.feedback_callback
        )

        self.send_goal_future.add_done_callback(self.goal_response_callback)

    def feedback_callback(self, feedback_msg) -> None:
        feedback: Fibonacci.Feedback = feedback_msg.feedback
        self.feedback_count += 1
        self.get_logger().info(
            f"Received feedback #{self.feedback_count}: {feedback.partial_sequence}"
        )

        if self.feedback_count > 0 and self.feedback_count >= self.cancle_after:
            self.get_logger().info("Requesting cancle...")
            self.goal_handle.cancel_goal_async()

    def goal_response_callback(self, future) -> None:
        goal_handle: ClientGoalHandle = future.result()
        if not goal_handle.accepted:
            self.get_logger().info("Goal rejected :(")
            return

        self.get_logger().info("Goal accepted :)")
        self.goal_handle = goal_handle

        self.get_result_future = goal_handle.get_result_async()
        self.get_result_future.add_done_callback(self.get_result_callback)

    def get_result_callback(self, future) -> None:
        result: Fibonacci.Result = future.result().result
        status = future.result().status
        self.get_logger().info(f"Final status: {status}")
        self.get_logger().info(f"Result: {result.sequence}")
        rclpy.shutdown()


def main(args: Optional[list[str]] = None) -> None:
    rclpy.init(args=args)
    action_client = FibonacciActionClient()
    action_client.send_goal(10, cancel_after=3)
    rclpy.spin(action_client)


if __name__ == "__main__":
    main()
